### ISSUE TITLE
Add support for mermaidjs state diagrams

### ISSUE THREAD

**Issue by tsurdilo** [serverlessworkflow/sdk-typescript#135]
**What would you like to be added**:
Simple support to spit out mermaid state diagram markup. Does not have to be anything fancy. 
https://mermaid-js.github.io/mermaid/#/stateDiagram

you can get idea from https://github.com/serverlessworkflow/sdk-java/tree/main/diagram which does similar thing just spits out plantuml markup

**Why is this needed**:
currently we have support in java sdk for https://plantuml.com/ and it works but requires us to deploy 
https://github.com/serverlessworkflow/workflow-diagram-service
on some cloud env, but we can't afford it as a project. 
This would allow us to use it in our vscode plugin as well

**Comment by antmendoza** [comment]
Hi,
I am working on this task,  

I have implemented already something and tested with this three examples and this is the output 

## [Application request](https://github.com/serverlessworkflow/specification/blob/main/examples/README.md#applicant-request-decision-example)

```

stateDiagram-v2
CheckApplication : CheckApplication
CheckApplication : type = switch

StartApplication : StartApplication
StartApplication : type = operation

RejectApplication : RejectApplication
RejectApplication : type = operation

[*] --> CheckApplication
CheckApplication --> StartApplication : ${ .applicants | .age >= 18 }
CheckApplication --> RejectApplication : ${ .applicants | .age < 18 }
StartApplication --> [*]
RejectApplication --> [*]

```

[![](https://mermaid.ink/img/eyJjb2RlIjoic3RhdGVEaWFncmFtLXYyXG5DaGVja0FwcGxpY2F0aW9uIDogQ2hlY2tBcHBsaWNhdGlvblxuQ2hlY2tBcHBsaWNhdGlvbiA6IHR5cGUgPSBzd2l0Y2hcblxuU3RhcnRBcHBsaWNhdGlvbiA6IFN0YXJ0QXBwbGljYXRpb25cblN0YXJ0QXBwbGljYXRpb24gOiB0eXBlID0gb3BlcmF0aW9uXG5cblJlamVjdEFwcGxpY2F0aW9uIDogUmVqZWN0QXBwbGljYXRpb25cblJlamVjdEFwcGxpY2F0aW9uIDogdHlwZSA9IG9wZXJhdGlvblxuXG5bKl0gLS0-IENoZWNrQXBwbGljYXRpb25cbkNoZWNrQXBwbGljYXRpb24gLS0-IFN0YXJ0QXBwbGljYXRpb24gOiAkeyAuYXBwbGljYW50cyB8IC5hZ2UgPj0gMTggfVxuQ2hlY2tBcHBsaWNhdGlvbiAtLT4gUmVqZWN0QXBwbGljYXRpb24gOiAkeyAuYXBwbGljYW50cyB8IC5hZ2UgPCAxOCB9XG5TdGFydEFwcGxpY2F0aW9uIC0tPiBbKl1cblJlamVjdEFwcGxpY2F0aW9uIC0tPiBbKl0iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/edit/##eyJjb2RlIjoic3RhdGVEaWFncmFtLXYyXG4gICAgV2hlbkNhcklzT24gOiBXaGVuQ2FySXNPblxuICAgIFdoZW5DYXJJc09uIDogdHlwZSA9IGV2ZW50XG4gICAgRG9DYXJWaXRhbENoZWNrcyA6IERvQ2FyVml0YWxDaGVja3NcbiAgICBEb0NhclZpdGFsQ2hlY2tzIDogdHlwZSA9IG9wZXJhdGlvblxuICAgIENoZWNrQ29udGludWVWaXRhbENoZWNrcyA6IENoZWNrQ29udGludWVWaXRhbENoZWNrc1xuICAgIENoZWNrQ29udGludWVWaXRhbENoZWNrcyA6IHR5cGUgPSBzd2l0Y2hcbiAgICBbKl0gLS0-IFdoZW5DYXJJc09uXG4gICAgV2hlbkNhcklzT24gLS0-IERvQ2FyVml0YWxDaGVja3NcbiAgICBEb0NhclZpdGFsQ2hlY2tzIC0tPiBDaGVja0NvbnRpbnVlVml0YWxDaGVja3NcbiAgICBDaGVja0NvbnRpbnVlVml0YWxDaGVja3MgLS0-IFsqXVxuIiwibWVybWFpZCI6IntcbiAgXCJ0aGVtZVwiOiBcImRlZmF1bHRcIlxufSIsInVwZGF0ZUVkaXRvciI6ZmFsc2UsImF1dG9TeW5jIjp0cnVlLCJ1cGRhdGVEaWFncmFtIjpmYWxzZX0)


## [Provision Orders](https://github.com/serverlessworkflow/specification/blob/main/examples/README.md#provision-orders-example)

```
stateDiagram-v2
ProvisionOrder : ProvisionOrder
ProvisionOrder : type = operation

MissingId : MissingId
MissingId : type = operation

MissingItem : MissingItem
MissingItem : type = operation

MissingQuantity : MissingQuantity
MissingQuantity : type = operation

ApplyOrder : ApplyOrder
ApplyOrder : type = operation

[*] --> ProvisionOrder
ProvisionOrder --> MissingId : Missing order id
ProvisionOrder --> MissingItem : Missing order item
ProvisionOrder --> MissingQuantity : Missing order quantity
ProvisionOrder --> ApplyOrder
MissingId --> [*]
MissingItem --> [*]
MissingQuantity --> [*]
ApplyOrder --> [*]
```

[![](https://mermaid.ink/img/eyJjb2RlIjoic3RhdGVEaWFncmFtLXYyXG5Qcm92aXNpb25PcmRlciA6IFByb3Zpc2lvbk9yZGVyXG5Qcm92aXNpb25PcmRlciA6IHR5cGUgPSBvcGVyYXRpb25cblxuTWlzc2luZ0lkIDogTWlzc2luZ0lkXG5NaXNzaW5nSWQgOiB0eXBlID0gb3BlcmF0aW9uXG5cbk1pc3NpbmdJdGVtIDogTWlzc2luZ0l0ZW1cbk1pc3NpbmdJdGVtIDogdHlwZSA9IG9wZXJhdGlvblxuXG5NaXNzaW5nUXVhbnRpdHkgOiBNaXNzaW5nUXVhbnRpdHlcbk1pc3NpbmdRdWFudGl0eSA6IHR5cGUgPSBvcGVyYXRpb25cblxuQXBwbHlPcmRlciA6IEFwcGx5T3JkZXJcbkFwcGx5T3JkZXIgOiB0eXBlID0gb3BlcmF0aW9uXG5cblsqXSAtLT4gUHJvdmlzaW9uT3JkZXJcblByb3Zpc2lvbk9yZGVyIC0tPiBNaXNzaW5nSWQgOiBNaXNzaW5nIG9yZGVyIGlkXG5Qcm92aXNpb25PcmRlciAtLT4gTWlzc2luZ0l0ZW0gOiBNaXNzaW5nIG9yZGVyIGl0ZW1cblByb3Zpc2lvbk9yZGVyIC0tPiBNaXNzaW5nUXVhbnRpdHkgOiBNaXNzaW5nIG9yZGVyIHF1YW50aXR5XG5Qcm92aXNpb25PcmRlciAtLT4gQXBwbHlPcmRlclxuTWlzc2luZ0lkIC0tPiBbKl1cbk1pc3NpbmdJdGVtIC0tPiBbKl1cbk1pc3NpbmdRdWFudGl0eSAtLT4gWypdXG5BcHBseU9yZGVyIC0tPiBbKl0iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/edit/##eyJjb2RlIjoic3RhdGVEaWFncmFtLXYyXG5DaGVja0FwcGxpY2F0aW9uIDogQ2hlY2tBcHBsaWNhdGlvblxuQ2hlY2tBcHBsaWNhdGlvbiA6IHR5cGUgPSBzd2l0Y2hcblxuU3RhcnRBcHBsaWNhdGlvbiA6IFN0YXJ0QXBwbGljYXRpb25cblN0YXJ0QXBwbGljYXRpb24gOiB0eXBlID0gb3BlcmF0aW9uXG5cblJlamVjdEFwcGxpY2F0aW9uIDogUmVqZWN0QXBwbGljYXRpb25cblJlamVjdEFwcGxpY2F0aW9uIDogdHlwZSA9IG9wZXJhdGlvblxuXG5bKl0gLS0-IENoZWNrQXBwbGljYXRpb25cbkNoZWNrQXBwbGljYXRpb24gLS0-IFN0YXJ0QXBwbGljYXRpb24gOiAkeyAuYXBwbGljYW50cyB8IC5hZ2UgPj0gMTggfVxuQ2hlY2tBcHBsaWNhdGlvbiAtLT4gUmVqZWN0QXBwbGljYXRpb24gOiAkeyAuYXBwbGljYW50cyB8IC5hZ2UgPCAxOCB9XG5TdGFydEFwcGxpY2F0aW9uIC0tPiBbKl1cblJlamVjdEFwcGxpY2F0aW9uIC0tPiBbKl0iLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)


## [Car Vitals Check](https://github.com/serverlessworkflow/specification/blob/main/examples/README.md#car-vitals-checks)

```
stateDiagram-v2
WhenCarIsOn : WhenCarIsOn
WhenCarIsOn : type = event

DoCarVitalChecks : DoCarVitalChecks
DoCarVitalChecks : type = operation

CheckContinueVitalChecks : CheckContinueVitalChecks
CheckContinueVitalChecks : type = switch

[*] --> WhenCarIsOn
WhenCarIsOn --> DoCarVitalChecks
DoCarVitalChecks --> CheckContinueVitalChecks
CheckContinueVitalChecks --> [*]

```

[![](https://mermaid.ink/img/eyJjb2RlIjoic3RhdGVEaWFncmFtLXYyXG5XaGVuQ2FySXNPbiA6IFdoZW5DYXJJc09uXG5XaGVuQ2FySXNPbiA6IHR5cGUgPSBldmVudFxuXG5Eb0NhclZpdGFsQ2hlY2tzIDogRG9DYXJWaXRhbENoZWNrc1xuRG9DYXJWaXRhbENoZWNrcyA6IHR5cGUgPSBvcGVyYXRpb25cblxuQ2hlY2tDb250aW51ZVZpdGFsQ2hlY2tzIDogQ2hlY2tDb250aW51ZVZpdGFsQ2hlY2tzXG5DaGVja0NvbnRpbnVlVml0YWxDaGVja3MgOiB0eXBlID0gc3dpdGNoXG5cblsqXSAtLT4gV2hlbkNhcklzT25cbldoZW5DYXJJc09uIC0tPiBEb0NhclZpdGFsQ2hlY2tzXG5Eb0NhclZpdGFsQ2hlY2tzIC0tPiBDaGVja0NvbnRpbnVlVml0YWxDaGVja3NcbkNoZWNrQ29udGludWVWaXRhbENoZWNrcyAtLT4gWypdXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/edit/##eyJjb2RlIjoic3RhdGVEaWFncmFtLXYyXG5Qcm92aXNpb25PcmRlciA6IFByb3Zpc2lvbk9yZGVyXG5Qcm92aXNpb25PcmRlciA6IHR5cGUgPSBvcGVyYXRpb25cblxuTWlzc2luZ0lkIDogTWlzc2luZ0lkXG5NaXNzaW5nSWQgOiB0eXBlID0gb3BlcmF0aW9uXG5cbk1pc3NpbmdJdGVtIDogTWlzc2luZ0l0ZW1cbk1pc3NpbmdJdGVtIDogdHlwZSA9IG9wZXJhdGlvblxuXG5NaXNzaW5nUXVhbnRpdHkgOiBNaXNzaW5nUXVhbnRpdHlcbk1pc3NpbmdRdWFudGl0eSA6IHR5cGUgPSBvcGVyYXRpb25cblxuQXBwbHlPcmRlciA6IEFwcGx5T3JkZXJcbkFwcGx5T3JkZXIgOiB0eXBlID0gb3BlcmF0aW9uXG5cblsqXSAtLT4gUHJvdmlzaW9uT3JkZXJcblByb3Zpc2lvbk9yZGVyIC0tPiBNaXNzaW5nSWQgOiBNaXNzaW5nIG9yZGVyIGlkXG5Qcm92aXNpb25PcmRlciAtLT4gTWlzc2luZ0l0ZW0gOiBNaXNzaW5nIG9yZGVyIGl0ZW1cblByb3Zpc2lvbk9yZGVyIC0tPiBNaXNzaW5nUXVhbnRpdHkgOiBNaXNzaW5nIG9yZGVyIHF1YW50aXR5XG5Qcm92aXNpb25PcmRlciAtLT4gQXBwbHlPcmRlclxuTWlzc2luZ0lkIC0tPiBbKl1cbk1pc3NpbmdJdGVtIC0tPiBbKl1cbk1pc3NpbmdRdWFudGl0eSAtLT4gWypdXG5BcHBseU9yZGVyIC0tPiBbKl0iLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)


In addition to changing the "type" description ("type = Event state" instead of "type = event" and so on..) 

can you think of any other changes, or any example that I can try in order to test the output ?

**Comment by tsurdilo** [comment]
@antmendoza holy crap ty!!!!!!! This is amazing. If you want to try some other samples you can try and compare the ones on 
serverlessworkflow.io/editor:

<img width="1385" alt="Screen Shot 2021-09-07 at 2 35 29 PM" src="https://user-images.githubusercontent.com/119422/132394343-6aeee685-03b8-4fb6-9af8-590b441b9fcc.png">


Do you think we should make this part of the typescript sdk or is there some way to externalize this into a nodejs app that uses the typescript sdk?

Awesome work man

**Comment by antmendoza** [comment]
thank you @tsurdilo 

> Do you think we should make this part of the typescript sdk or is there some way to externalize this into a nodejs app that uses the typescript sdk?

I think that ideally this should be a different project, but on the other hand the implementation is [only 2 or 3 files](https://github.com/antmendoza/sdk-typescript/tree/mermaid/src/lib/diagram), so I don't know if really makes sense to maintain a different project for now.

**Comment by antmendoza** [comment]
"only few files" for now, let's see if I can make this work and we can decide later.

**Comment by tsurdilo** [comment]
i think as long if its somehow possible to use it on our /editor page thats not a nodejs project (just html and js) it does not matter in the end. wdyt?

**Comment by tsurdilo** [comment]
i also think we should use mermaidjs output in java sdk too so they are consistent. will look at your impl to copy :)

**Comment by antmendoza** [comment]
what I have implemented here is a class that receives a workflow parameter and generate the mermaid source code. (not the svg)
it returns something like 

```
stateDiagram-v2
CheckApplication : CheckApplication
CheckApplication : type = switch

StartApplication : StartApplication
StartApplication : type = operation

RejectApplication : RejectApplication
RejectApplication : type = operation

[*] --> CheckApplication
CheckApplication --> StartApplication : ${ .applicants | .age >= 18 }
CheckApplication --> RejectApplication : ${ .applicants | .age < 18 }
StartApplication --> [*]
RejectApplication --> [*]
```

mermaidjs relies on the browser to render the image, so the actual implementation is enough for the online editor (I have already tested it), but, do we want to expose in our sdk a class called `mermaidCodeGenerator` or something similar? 

I have been testing the implementation to generate an image directly in the sdk, to make the API more generic, but with this implementation I've realised that the sdk is not working in the browser.

to generate the image in the sdk I've had to use a [ headless browser (or mermaid-cli that at the end is using a headless browser )](https://github.com/mermaid-js/mermaid/issues/593), and by doing that, for some reason, the final bundle is incompatible with the browser. I need to investigate. 

So if you think that generating the mermaid source code is enough for now,  and we need this for the editor, I can finish the actual implementation and later continue working on this and we will see how we proceed.
